### PR TITLE
Add withdrawn list

### PR DESCRIPTION
**Pull request by BernardHan**

### CODE REVIEW

File: src/db/index.js [hunk 1]
```diff
@@ -113,27 +114,36 @@ async function applyMigrations() {
...
+  async.whilst(
+    condition => condition(null, i < migrations.length),
+    (callback) => {
+      const migration = migrations[i];
+      i++;
+      try {
+        if (Number(migration.number) === lastMigration + 1) {
+          // Run migration
+          logger.info(`Running migration ${migration.number}...`);
+          migration.migrate(db);
```
Comment (dwalintukan) [inline src/db/index.js]: need to await here? the migration func is async.

File: src/db/index.js [hunk 2]
```diff
@@ -113,27 +114,36 @@ async function applyMigrations() {
...
-  }
-  /* eslint-enable no-restricted-syntax, no-await-in-loop */
-
-  logger.info('Migrations complete.');
+      callback(null, i);
+    },
+    (err, number) => {
+      if (err) {
+        logger.error(`Migration ${number} error: ${err.message}`);
+        return;
```
Comment (dwalintukan) [inline src/db/index.js]: throw here instead of return. we cant continue if a migration failed.

File: src/db/index.js [hunk 3]
```diff
@@ -113,27 +114,36 @@ async function applyMigrations() {
...
+      try {
+        if (Number(migration.number) === lastMigration + 1) {
+          // Run migration
+          logger.info(`Running migration ${migration.number}...`);
+          migration.migrate(db);
+          logger.info(`Migration ${migration.number} done`);
+          // Track the last migration number
+          lastMigration = migration.number;
+          fs.outputFileSync(migrationTrackPath, `LAST_MIGRATION=${migration.number}\n`);
+          logger.info('Write file done');
```
Comment (dwalintukan) [inline src/db/index.js]: nit: more descriptive log, "migration.dat updated"?

File: src/db/index.js [hunk 4]
```diff
@@ -113,27 +114,36 @@ async function applyMigrations() {
...
+        }
+      } catch (err) {
+        callback(err, migration.number);
```
Comment (dwalintukan) [inline src/db/index.js]: The loop will stop here cause the `err != null` right?

File: src/db/migrations/migration1.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      const withdraw = withdraws[i];
+      let withdrawnList = [];
+      if (withdrawnLists.has(withdraw.eventAddress)) {
+        withdrawnList = withdrawnLists.get(withdraw.eventAddress);
+      }
+      withdrawnList.push(withdraw.winnerAddress);
+      withdrawnLists.set(withdraw.eventAddress, withdrawnList);
+    }
+
+    withdrawnLists.forEach((value, key) => {
```
Comment (dwalintukan) [inline src/db/migrations/migration1.js]: Need to create a list of promises, add all the update promises, then call `await Promise.all`. This current logic wont wait for all the db updates to be done before continuing.

File: src/db/migrations/migration1.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      DBHelper.updateEventByAddress(
```
Comment (dwalintukan) [inline src/db/migrations/migration1.js]: await

File: src/db/db-helper.js [hunk 7]
```diff
@@ -206,6 +216,24 @@ module.exports = class DBHelper {
     }
   }
 
+  static async updateEventWithdrawnList(withdraw) {
+    try {
+      const event = await DBHelper.findOneEvent({ address: withdraw.eventAddress });
+      if (isNull(event)) throw Error('Event does not exist');
```
[the remark this answers is not in the document]
Comment (BernardHan) [reply src/db/db-helper.js]: The func would return at the throw I think?

File: src/db/index.js (hunk 1 above)
Comment (BernardHan) [reply src/db/index.js]: Looks like it doesn't allow await, only callback

File: src/db/index.js (hunk 4 above)
Comment (BernardHan) [reply src/db/index.js]: right

**Comment by BernardHan** [comment]
@dwalintukan I have rewrote the apply migration, and made the changes

I made detailed comments to explain what's happening.

**Pull request merged**